components/dropdown/dropdown-popover-mixin.js

```diff
@@ -0,0 +1,342 @@
+			super.configure({
+				maxHeight: this.maxHeight,
+				maxWidth: this.maxWidth,
+				minHeight: this.minHeight,
+				minWidth: this.minWidth,
+				mobileBreakpoint: this.mobileBreakpointOverride,
+				mobileTrayLocation: this.#adaptMobileTrayLocation(this.mobileTray),
+				noAutoClose: this.noAutoClose,
+				noAutoFit: this.noAutoFit,
+				noAutoFocus: this.noAutoFocus,
+				noPointer: this.noPointer,
+				offset: (this.verticalOffset !== undefined ? Number.parseInt(this.verticalOffset) : undefined),
+				position: { location: 'block-end', span: this.#adaptPositionSpan(this.align) },
+				trapFocus: this.trapFocus
+			});
+		}
+
+		if (changedProperties.has('opened')) {
+			if (this.opened) this.open(true);
+			else if (changedProperties.get('opened')) this.close();
+		}
+	}
+
+	#contentElement;
+
+	#adaptMobileTrayLocation(val) {
+		switch (val) {
+			case 'bottom': return 'block-end';
+			case 'left': return 'inline-start';
+			case 'right': return 'inline-end';
+			default: return undefined;
+		  }
+	}
+
+	#adaptPositionSpan(val) {
+		switch (val) {
+			case 'start': return 'end';
+			case 'end': return 'start';
+			default: return 'all';
+		  }
+	}
+
+	#getMobileCloseButtonStyles() {
+		if (!this._mobile || !this._mobileTrayLocation) {
+			return { display: 'none' };
+		}
+
+		let footerWidth;
+		if (this.noPaddingFooter) {
+			footerWidth = 'calc(100% - 24px)';
+		} else if (this._hasFooterSlotContent) {
+			footerWidth = '100%';
+		} else {
+			footerWidth = 'calc(100% + 16px)';
+		}
+
+		return {
+			display: !this.noMobileCloseButton ? 'inline-block' : 'none',
+			marginBlock: this._hasFooterSlotContent ? '0' : '-20px -20px',
+			marginInline: this._hasFooterSlotContent ? '0' : '-20px 0',
+			padding: this._hasFooterSlotContent && !this.noPaddingFooter ? '12px 0 0 0' : '12px',
+			width: footerWidth
+		};
+	}
+
+	#handleFooterSlotChange(e) {
+		this._hasFooterSlotContent = e.target.assignedNodes().length !== 0;
+	}
+
+	#handleHeaderSlotChange(e) {
+		this._hasHeaderSlotContent = e.target.assignedNodes().length !== 0;
+	}
+
+	#handlePopoverClose() {
+		this.opened = false;
+
+		/** Dispatched when the dropdown is closed */
+		this.dispatchEvent(new CustomEvent('d2l-dropdown-close', { bubbles: true, composed: true }));
+	}
+
+	#handlePopoverOpen() {
+		this.opened = true;
+
+		if (!this.noAutoFit && this.#contentElement) {
+			this.#contentElement.scrollTop ??= 0;
+		}
+
+		/** Dispatched when the dropdown is opened */
+		this.dispatchEvent(new CustomEvent('d2l-dropdown-open', { bubbles: true, composed: true }));
+	}
+
+	#handlePopoverPosition() {
+		this.#toggleScrollStyles();
+
+		/** Dispatched when the dropdown position finishes adjusting */
+		this.dispatchEvent(new CustomEvent('d2l-dropdown-position', { bubbles: true, composed: true }));
+	}
+
+	#toggleScrollStyles() {
+		this._blockEndScroll = this.#contentElement.scrollHeight - (this.#contentElement.scrollTop + this.#contentElement.clientHeight) >= 5;
```

Comment:
In the back of my mind I have the idea of beginning to use the scroll-state container queries, but I felt moving to that should be kept separate from these changes.